net/captivedetection/captivedetection.go

```diff
@@ -0,0 +1,226 @@
+}
+
+func isDefinitelyNotWiFi(ifName string) bool {
+	return strings.HasPrefix(ifName, "tailscale") ||
+		strings.HasPrefix(ifName, "utun") ||
+		strings.HasPrefix(ifName, "awdl") ||
+		strings.HasPrefix(ifName, "bridge") ||
+		strings.HasPrefix(ifName, "ap") ||
+		strings.HasPrefix(ifName, "tun") ||
+		strings.HasPrefix(ifName, "tap")
+}
+
+// checkCaptivePortal reports whether or not we think the system is behind a
+// captive portal, detected by making a request to a URL that we know should
+// return a "204 No Content" response and checking if that's what we get.
+//
+// The boolean return is whether we think we have a captive portal.
+func checkCaptivePortal(ctx context.Context, ifIndex int, preferredDERPRegionID int, logf func(format string, a ...any)) bool {
+	defer httpClient.CloseIdleConnections()
+
+	endpoints := availableEndpoints(preferredDERPRegionID)
+	if len(endpoints) == 0 {
+		logf("no endpoints available for captive portal detection")
+		return false
+	}
+
+	// To avoid false positives, we issue requests to a maximum of 6 captive portal detection URLs
+	// and only report a captive portal is present when at least 2 responses look like a captive
+	// portal.
+	foundCount := 0
+	for _, e := range endpoints[:6] {
+		found, err := checkCaptivePortalEndpoint(ctx, e, ifIndex, logf)
+		if err != nil {
+			logf("checkCaptivePortalEndpoint failed with endpoint %s: %v", e.URL, err)
+			continue
+		}
+		if found {
+			foundCount++
+		}
+		if foundCount == 2 {
+			return true
+		}
+	}
+	return false
+}
+
+// checkCaptivePortalEndpoint checks if the given Endpoint is a captive portal by making an HTTP request to the
+// given Endpoint URL and checking if the response looks like a captive portal.
+func checkCaptivePortalEndpoint(ctx context.Context, e Endpoint, ifIndex int, logf func(format string, a ...any)) (found bool, err error) {
+	req, err := http.NewRequestWithContext(ctx, "GET", e.URL, nil)
+	if err != nil {
+		return false, err
+	}
+
+	// Attach the Tailscale challenge header if the endpoint supports it. Not all captive portal detection endpoints
+	// support this, so we only attach it if the endpoint does.
+	if e.SupportsTailscaleChallenge {
+		if u, err := url.Parse(e.URL); err == nil {
+			// Note: the set of valid characters in a challenge and the total
+			// length is limited; see isChallengeChar in cmd/derper for more
+			// details.
+			chal := "ts_" + u.Host
+			req.Header.Set("X-Tailscale-Challenge", chal)
+		}
+	}
+
+	// Force the HTTP connection to use the given interface.
+	// Why? On Darwin, until you've accepted or closed the macOS captive portal alert,
+	// no default route interface is available, so all requests will fail unless we do this.
+	// TODO: investigate if this is also necessary on other platforms (Windows?)
+	d := &net.Dialer{
+		Control: func(network, address string, c syscall.RawConn) error {
+			return prepareRawConn(c, ifIndex, logf)
+		},
+	}
+	httpClient.Transport = &http.Transport{
```

Comment:
This isn't concurrency-safe; if multiple callers call `DetectCaptivePortal` at the same time, this will be a data race. You could move the `httpClient` into a type, or add a mutex, or copy the client here. Since `http.Client` doesn't contain private fields, you can shallow-copy it:
```go
localHTTPClient := ptr.To(*httpClient)
```

Reply:
Added a Detector type which has its own HTTP Client, and is locked by a mutex so that only one detection attempt can run at a time.